Use the fiscal start day of the month for relative quarter reports

agwekixj opened the pull request:
Only the quarter's start month in reports is adjusted when Absolute is selected in the Start Date on the Accounting Period preferences. This branch adjusts both the quarter's start day and month when Absolute is selected or when Relative and Today, Start of this month, or Start of previous month are selected.

agwekixj committed:
Use the fiscal start day of the month for relative quarter reports

Only the quarter's start month in reports is adjusted when Absolute is selected in the Start Date on the Accounting Period preferences. This branch adjusts both the quarter's start day and month when Absolute is selected or when Relative and Today, Start of this month, or Start of previous month are selected.

agwekixj commented:
The test errors are expected as we're changing the prior relative quarter start behavior.

jralls commented:
> The test errors are expected as we're changing the prior relative quarter start behavior.

In only one place, so that the reports' understanding of the relative quarter differs from the rest of the program's. This is particularly amusing because it answers my "what am I missing" from #2064: In the absence of preferences, which is the case with unit tests, the default accounting period is `TODAY`.

That aside it appears from grepping around that the only other place that would be affected by this policy change is in the price database, where leaving the last quote of each quarter is an option for deleting old quotes. That's not exactly high-risk.

However while I don't advocate as extensive a testing range as [Graham Menhemnitt](https://bugs.gnucash.org/show_bug.cgi?id=798946#c31) it does deserve more testing, not less.

agwekixj commented:
I have updated the expected test results for the new relative quarter start behavior.

agwekixj commented:
The end of a quarter should always be one day before the start of the next quarter.

jralls committed:
Add and use a fake preferences backend to set the fiscal year.

This restores the expected precondition for the quarter tests of the
Accounting Period being the current calendar year, replacing the
today-relative quarter begin and end dates.

jralls committed:
Test gnc-option-date variable fy start days.

Leverages the FakePrefsBackend from the previous commit to change the
fiscal year for testing.

Makes now_t a defaulted parameter of gnc_relative_date_to_time64 so
that we can inject different reference dates for testing.

jralls commented:
@agwekixj Don't merge stable into a PR, always rebase the PR on stable.

I did that and force-pushed.

I've also replaced your today-relative quarters hack with a fake prefs backend that sets the FY so that the original tests work correctly and leveraged it to write new tests for the fy-relative quarter values.

Thanks for working on this.

agwekixj commented on libgnucash/engine/gnc-option-date.cpp:
```diff
@@ -525,10 +525,9 @@ gnc_relative_date_to_time64(RelativeDatePeriod period)
         return gnc_accounting_period_fiscal_end();
 
-    GncDateTime now_t;
     if (period == RelativeDatePeriod::TODAY)
```
This condition is already handled.

jralls replied:
I don't understand the comment. The change you've highlighted is part of making `now_t` a parameter of the function.

agwekixj replied:
Examine the condition handled on the first two lines of the function (521 - 522) and then on lines (528 - 529).

jralls replied:
Got it, thanks.
